### PR TITLE
[Distributed] Pytorch example of distributed GraphSage.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
This converts GraphSage to distributed training.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/graphsage/train_dist.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    parser.add_argument('--local_rank', type=int, help='get rank of the process')
+    parser.add_argument('--num_procs', type=int, help='the number of training processes')
+    args = parser.parse_args()
+
+    print(args)
+
+    if args.server:
+        start_server(args)
+    else:
+        main(args)
```
Comment (BarclayII) [inline examples/pytorch/graphsage/train_dist.py:187]: Can we have `start_server` and `main` as two separate scripts?

**Comment by zheng-da** [comment]
@BarclayII this isn't the full user experience. We'll provide a launch script that sets up everything so users only need to partition a graph, move data to the right location and use a launch script to run the training job.

**Comment by BarclayII** [comment]
OK.  Can we have it in a separate directory like `_experimental` then because this is not yet the full experience?

**Comment by zheng-da** [comment]
sure. to answer your question:
2. just upload ip_config.txt
3. every machine needs to access ip_config.txt and partitioned data. one option is to copy them to all machines. the better option is to copy them to global storage such as NFS. when we have a launch script, ip_config.txt only needs to be in the master node and partitioned data need to be in global storage such as NFS, S3, HDFS, etc.
4. I'll provide a multi-machine example.

File: examples/pytorch/graphsage/experimental/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+export PYTHONPATH=$PYTHONPATH:..
+```
+
+In this example, we partition the OGB product graph into 4 parts with Metis. The partitions are balanced with respect to
+the number of nodes, the number of edges and the number of labelled nodes.
+```bash
+# partition graph
+python3 partition_graph.py --dataset ogb-product --num_parts 4 --balance_train --balance_edges
+```
+
```
Comment (aksnzhy) [inline examples/pytorch/graphsage/experimental/README.md:21]: Should we put the `--banlace_train` and `--balance_edges` as default arguments?

File: examples/pytorch/graphsage/experimental/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+python3 train_dist.py --server --graph-name ogb-product --id 3 --num-client 4 --conf_path data/ogb-product.json --ip_config ip_config.txt
+```
+
+### Step 4: run trainers
+We run a trainer process on each machine. Here we use Pytorch distributed. We need to use pytorch distributed launch to run each trainer process.
+Pytorch distributed requires one of the trainer process to be the master. Here we use the first machine to run the master process.
+
+```bash
+# run client on machine 0
+python3 -m torch.distributed.launch --nproc_per_node=1 --nnodes=4 --node_rank=0 --master_addr="172.31.16.250" --master_port=1234 train_dist.py --graph-name ogb-product --ip_config ip_config.txt --num-epochs 3 --num-client 4 --batch-size 1000 --lr 0.1
```
Comment (aksnzhy) [inline examples/pytorch/graphsage/experimental/README.md:48]: Does node_rank is necessary? `kvstore.machine_id` can provide this information.

Comment (aksnzhy) [reply examples/pytorch/graphsage/experimental/README.md:48]: `--num_client` can be removed after the new pr for rpc is merged.

Comment (aksnzhy) [reply examples/pytorch/graphsage/experimental/README.md:48]: Why we need a `--master_addr`?

Comment (zheng-da) [reply examples/pytorch/graphsage/experimental/README.md:48]: these parameters are used by pytorch distributed. they are not used by DGL.

File: examples/pytorch/graphsage/experimental/README.md (hunk 2 above)
Comment (zheng-da) [reply examples/pytorch/graphsage/experimental/README.md:21]: sure. although this is just an example.

**Pull request merged**